lesson-3-sheduling-all-around/README.md

```diff
@@ -0,0 +1,232 @@
+            - S1
+        topologyKey: kubernetes.io/hostname
+  containers:
+  - name: pod-antiaffinity
+    image: docker.io/ocpqe/hello-pod
+```
+ 
+Now you can use `oc` command to doublecheck if everything is running correctly.
+
+``` bash
+oc get pods -o jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.spec.nodeName}{"\n"}'
+```
+
+As a task for yourself please create pod "p4" which will be scheduled on same node as pod "p3".
+You can also expand jsonpath filter to print the value of "security" label.
+
+If you are interested more in jsonpath you can find the docs at https://kubernetes.io/docs/reference/kubectl/jsonpath/
+
+
+To cleanup the environment we will run
+``` bash
+oc delete pod p1 p2 p3 p4
+```
+
+You can use these selector for more stuff. If you propagate your server placements, availability zones or HW equipment to lables. But be aware of bad resource utilization which can come with a complex rules and can even end up with upgrade issues.
+
+Also think about scalability. For example if you need two services to be running on the same node consider them being part of one pod - then they can scale easily together.
+
+Quality of Service
+==================
+
+One of the part of workload schediling is also working with Quality of Service to assure that we have enough resources
+to fullfill the needs of our workload.
+
+TBD
+
+
+Custome Scheduler
+=================
+
+Custom scheduler is a little bit overhead in a lot of cases. We are using it to prove that it is possible and to demonstrate how to access k8s api from pods. The complete source of the scheduler can be viewed in the `scheduler/app.py`. In this scenario we will:
+
+- setup OpenShift build for the scheduler app
+- build it
+- create service account and set th required permissions
+- run scheduler in OpenShift pod
+- Schedule a new pod using this scheduler
+
+
+So let's start with creating a binary build. This will enable us to build container in OpenShift using files on local filesystem. So run:
+``` bash
+oc new-build --binary --name=sched
+```
+
+
+
+
+Then we ask OpenShift to build our container by uploading the contet of a "scheduler" directory. Before executing this please change "NAME_SPACE" text to the project/namespace you are running your pods in. And then execute:
+
+``` bash
+oc start-build sched --from-dir scheduler
+```
+
+we need to wait for build to succeed - we can check it the status by running
+``` bash
+oc get builds -w
+```
+
+to see the log of a build you can run
+``` bash
+oc logs sched-1-build --follow
+```
+
+When the job in the state of "Complete" we can continue by creating the service account. IT doesn't have enough permissions - but we will fix this in next step by meking it cluster-admin.
+
+``` bash
+oc create sa sched
+```
+
+Now please ask in google meet chat for adding permissions to your service account. If you are on your cluster you can do it by command:
+
+``` bash
+oc adm policy add-cluster-role-to-user cluster-admin -z sched
```

Comment:
ohoho, we're going very bold in this demo

Reply:
exactly my thoughts! :O

Reply:
Do we need cluster-admin, btw? It opens up doors to the nodes and possibly to the infrastructure running the cluster :eyes: